### PR TITLE
[build] A build tool manager

### PR DESCRIPTION
**Pull request by pkriens**
A tool to update the workspace root with a build
tool like gradle, see buildtool.md for a
description.

Signed-off-by: Peter Kriens <Peter.Kriens@aqute.biz>

### CODE REVIEW

File: biz.aQute.bnd/src/aQute/bnd/main/bnd.java [hunk 1]
```diff
@@ -4583,4 +4585,63 @@ private File toFile(Container c) {
...
+				getInfo(toolmanager);
+			}
+		}
+		getInfo(ws);
+	}
+
+	private String getUrl(String key, Attrs attrs) {
+		String version = attrs.getVersion();
+		if (Verifier.isVersion(version)) {
+			return "https://github.com/bndtools/workspace.tool." + key + "/archive/refs/tags/" + version + ".zip";
```
Comment (bjhargrave) [inline biz.aQute.bnd/src/aQute/bnd/main/bnd.java:4622]: Is this a collection of new repos? `bndtools/workspace.tool.?`. Who will be maintaining these repos?

File: biz.aQute.bndlib/src/aQute/bnd/buildtool/ToolManager.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+	public String install(String url, Map<String, String> parameters, boolean force) {
+		HttpClient client = workspace.getPlugin(HttpClient.class);
+		try {
+			if (client.isOffline()) {
+				return "The workspace is off line";
+			}
+
+			if (Verifier.isVersion(url)) {
+				aQute.bnd.version.Version v = new aQute.bnd.version.Version(url);
+				url = "https://github.com/bndtools/gradle.workspace/archive/refs/tags/" + v.toStringWithoutQualifier()
```
Comment (bjhargrave) [inline biz.aQute.bndlib/src/aQute/bnd/buildtool/ToolManager.java:55]: This is another new repo? It does not match the pattern above `bndtools/workspace.tool.?`.

File: docs/_instructions/buildtool.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
+---
+layout: default
+class: bnd
+title: -buildtool  toolspec  (EXPERIMENTAL!)
```
Comment (bjhargrave) [inline docs/_instructions/buildtool.md:4]: How does this related to the templates in Bndtools which setup a new workspace? Should they be kept in sync/use the same source repo?

File: biz.aQute.bndlib/src/aQute/bnd/buildtool/ToolManager.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+									.append(e.getValue())
+									.append("\n");
+							}
+							resource = new EmbeddedResource(sb.toString()
+								.getBytes(StandardCharsets.UTF_8), 0);
+						}
+
+						IO.copy(resource.openInputStream(), destination);
+
+						if (attrs.containsKey("exec")) {
```
Comment (bjhargrave) [inline biz.aQute.bndlib/src/aQute/bnd/buildtool/ToolManager.java:143]: What if it is `exec=false`?

File: biz.aQute.bndlib/src/aQute/bnd/buildtool/ToolManager.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+						destination.getParentFile()
+							.mkdirs();
+						if (attrs.containsKey("macro")) {
+							actions += " macro";
+							String content = IO.collect(resource.openInputStream());
+							content = getReplacer().process(content);
+							resource = new EmbeddedResource(content.getBytes(StandardCharsets.UTF_8), 0);
+						}
+						if (attrs.containsKey("append")) {
```
Comment (bjhargrave) [inline biz.aQute.bndlib/src/aQute/bnd/buildtool/ToolManager.java:126]: What if it is `append=false`?

File: biz.aQute.bndlib/src/aQute/bnd/buildtool/ToolManager.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+						if (destination.isFile()) {
+							if (!force)
+								return "cannot overwrite " + destination;
+
+							actions += " overwrite";
+						}
+
+						destination.getParentFile()
+							.mkdirs();
+						if (attrs.containsKey("macro")) {
```
Comment (bjhargrave) [inline biz.aQute.bndlib/src/aQute/bnd/buildtool/ToolManager.java:120]: What if it is `macro=false`?

File: biz.aQute.bndlib/src/aQute/bnd/buildtool/ToolManager.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+						progress("not included %s", path);
+						continue;
+					}
+
+					String newPath = Strings.stripPrefix(path, prefix);
+					Instruction matcher = copyInstructions.matcher(newPath);
+					Attrs attrs = copyInstructions.get(matcher);
+					String actions = newPath;
+
+					if (attrs.containsKey("skip")) {
```
Comment (bjhargrave) [inline biz.aQute.bndlib/src/aQute/bnd/buildtool/ToolManager.java:106]: What if it is `skip=false`?

File: biz.aQute.bnd/src/aQute/bnd/main/bnd.java (hunk 1 above)
Comment (pkriens) [reply biz.aQute.bnd/src/aQute/bnd/main/bnd.java:4622]: You already maintain this for the workspace, so I thought of making a Github action?

File: biz.aQute.bndlib/src/aQute/bnd/buildtool/ToolManager.java (hunk 2 above)
Comment (pkriens) [reply biz.aQute.bndlib/src/aQute/bnd/buildtool/ToolManager.java:55]: mistake, left over, dead code

File: docs/_instructions/buildtool.md (hunk 3 above)
Comment (pkriens) [reply docs/_instructions/buildtool.md:4]: first see how this works out, I am tired of all the time having to copy these gradle files. So I am trying to create a clean workspace and then use this tool to install gradle (or another tool) with a Github action. If it works well, we should take a good look at our workspaces and reduce. but one step at a time.

**Pull request merged**